### ISSUE TITLE
Projen python version default command fails

### ISSUE THREAD

**Issue by floydspace** [projen/projen#2103]
Hello, guys, thank you for the awesome tool, the nodejs version works perfectly, however, I'm struggling with the Python binary. Simply installed it globally using `pipx install projen` and running `projen`, fails with the following error:
```sh
➜  ~ projen
Traceback (most recent call last):
  File "/Users/victor/.local/bin/projen", line 10, in <module>
    __jsii_assembly__.invokeBinScript("projen", "projen", sys.argv[1:])
  File "/Users/victor/.local/pipx/venvs/projen/lib/python3.10/site-packages/jsii/_runtime.py", line 52, in invokeBinScript
    response = _kernel.invokeBinScript(pkgname, script, args)
  File "/Users/victor/.local/pipx/venvs/projen/lib/python3.10/site-packages/jsii/_kernel/__init__.py", line 312, in invokeBinScript
    args=_make_reference_for_native(self, args),
  File "/Users/victor/.local/pipx/venvs/projen/lib/python3.10/site-packages/jsii/_kernel/__init__.py", line 204, in _make_reference_for_native
    kernel.create(d.__class__, d)
  File "/Users/victor/.local/pipx/venvs/projen/lib/python3.10/site-packages/jsii/_kernel/__init__.py", line 323, in create
    fqn=klass.__jsii_type__ or "Object",
AttributeError: type object 'tuple' has no attribute '__jsii_type__'
```
I tried py3.8 and py3.10, and encounter the same result. weird that there is no such issue yet, I'm the only one having this error?
Thank you for your help.

**Comment by mrgrain** [comment]
Hi @floydspace ! I'm not actually sure this is supposed to work or even possible to work with jsii.

Background: Projen is a TypeScript project with [jsii bindings](https://aws.github.io/jsii/) for Python and other languages so that you can write the configuration in the language of your project (or choice). Due to its nature jsii has limitations and running a cli might just be one of them.

tl;dr the expected way to run projen is still via node: `npx projen`

--- 

I appreciate we're probably not doing the best job explaining this in the documentation.

**Comment by floydspace** [comment]
Thank you for the clarification.

> the expected way to run projen is still via node: `npx projen`

yes, this is what I do.

However, in my python environment, I feel like using python tools and not being dependent on nodejs environment. I know eventually implicitly I still depend on it, but better implicitly than explicitly.

Anyway, I think it's still a bug for python binary, but decide your self should the issue be opened or closed, depends on your plans regarding the projen version for python

**Comment by mrgrain** [comment]
> However, in my python environment, I feel like using python tools and not being dependent on nodejs environment. I know eventually implicitly I still depend on it, but better implicitly than explicitly.

I wholeheartedly with this sentiment. Binaries for all languages or a standalone binary should be a long-term goal.


> Anyway, I think it's still a bug for python binary, but decide your self should the issue be opened or closed, depends on your plans regarding the projen version for python

We rely on jsii to provide cross-lang capabilities. I can see that we announce projen as an app, so `pipx` tries to run it. What I'm currently not sure about is if that's supposed to happen (and we need to fix a bug with running it as an app) or if announcing an app was accidental (and we should stop doing that). If it's the latter, a long term fix would depend on jsii's capabilities in that area, or we start shipping  a standalone binary.

**Comment by jmalins** [comment]
I was encountering the same error. It actually looks like it is tied to this change (line 52):
https://github.com/aws/jsii/pull/3694/files#diff-8c3c3ced74aeb038ac90c893f7998ce488d3771561a5d253376f9177bcef707c

The arguments for the python script are not being marshaled correctly.

I corrected it in my locally deployed install, and got to the point of finding another error: a missing `+x` permission in the JSII tarball bundled with python package. Once that was (manually) fixed, the `projen` executable worked as expected.

I'd like to make a PR, but would still need to navigate the source to figure out where the `projen` python script is defined and where the tarball is created.

**Comment by jmalins** [comment]
To clarify, the issues are in a combination of JSII (argument marshaling) and projen (syntax of the `projen` app that is installed in the venv).

At this point, I am not clear on why the `+x` permission is dropped on JSII tarball that is distributed along with projen python package... that could be either a JSII issue or projen toolchain/release issue. I need to do more research.

**Comment by mrgrain** [comment]
@RomainMuller might have insights on the jsii side of things

**Comment by jmalins** [comment]
`jsii` 1.69.0 with the fix above landed. Unfortunately, the default `projen` script is still has problems.

It appears that the JSII tarball included with the python package is missing `+x` permissions on the `bin/projen` executable.
```sh
# tar tvf .env/lib/python3.9/site-packages/projen/_jsii/projen@0.62.25.jsii.tgz | grep "package/bin/"
-rw-r--r--  0 0      0         118 Oct 26  1985 package/bin/projen
```
If I re-write the tarball to have an executable `projen` script, everything works properly.

Running the `package:python` target locally produces a `dist/python` Twine package that contain a JSII tarball with the correct permissions. I'm at a loss for why the published artifact is different. On CI, there is an additional `rsync -a` step, but the man page seems to suggest this will preserve permissions.

Any thoughts?

**Comment by mrgrain** [comment]
After some research, it might be related to this issue:
https://github.com/actions/upload-artifact/issues/38

**Comment by mrgrain** [comment]
Workaround documented here:
https://github.com/actions/upload-artifact#maintaining-file-permissions-and-case-sensitive-files

**Comment by jmalins** [comment]
I agree, this does seem to be the issue.

Is the best solution to add tar/untar to `build/build-workflow.ts`? This is the likely the simplest solution, but it would be a behavior change to the uploaded artifact. An alternative would be to backup and restore the permissions, similar to how the self-mutation patch is handled.

**Comment by mrgrain** [comment]
I'm not sure to be honest. We might have to play around with options to see which we hate the least?

- Tar would be generic, but I doubt the current code is setup for it very well. I.e. you will have to change it in a lot of places
- Generic Permissions restore could work, but maybe it's a bite brittle? What did you have in mind for it?
- We know that bin files [listed here](https://github.com/projen/projen/blob/main/package.json#L8-L10) definitely need `+x`. Maybe a task that does that and we can run it before packaging?

**Comment by jmalins** [comment]
Explicitly setting permissions on just the bin scripts is appealing. It will fix the problem at hand and be simple to implement.

To me, the main argument against that approach is that it leaves a subtle landmine in every user-space `JsiiProject`. I suppose one could argue that there is no use case for arbitrary executables in a `JsiiProject`, other than bin scripts, so maybe the chance of stepping on it is low. 

For backup and restore of permissions, I was considering `getfacl`/`setfacl`, as described [here](https://www.linux.com/training-tutorials/how-easily-back-and-restore-linux-file-permissions/). A quick test confirmed that they are present on GitHub's `ubuntu-latest` image. This would also be simple to implement, but I'm not sure if there would be unintended side effects without testing.

**Comment by mrgrain** [comment]
> To me, the main argument against that approach is that it leaves a subtle landmine in every user-space `JsiiProject`.

I fully agree with that. The best fix would be for the GitHub Action to support permissions, but looks like that's not going to happen. Saying that, it appears to be the first time an issue with permissions came up.

> Explicitly setting permissions on just the bin scripts is appealing. It will fix the problem at hand and be simple to implement.
For backup and restore of permissions, I was considering `getfacl`/`setfacl`.  This would also be simple to implement, but I'm not sure if there would be unintended side effects without testing.

I'm probably okay with either approach. For `getfacl`/`setfacl` you might be able to limit the impact by only applying it to `.repo`

**Comment by jmalins** [comment]
Success! This now works:
```sh
~> cd my-project
~/my-project> npx projen new awscdk-app-py
... #install output here
~/my-project> source .env/bin/activate.fish
(.env) ~/my-project> projen --version
0.62.28
```
Python users can now bootstrap a new project and interact with a local projen instance without having to install anything globally using the `.env/bin/projen` script that is included in the venv path.

**Comment by mrgrain** [comment]
And `pipx run projen new awscdk-app-py` also works! This is great. Good work everyone and especially @jmalins 🥳

**Comment by floydspace** [comment]
@jmalins thank you very much, I'm impressed. It works!!

**Comment by mrgrain** [comment]
Does either of you have the skills handy to test if this also works with the Go and Java packages?

**Comment by jmalins** [comment]
Not handy, unfortunately.
